### PR TITLE
Implementation of Geth node debugging API's (web3-eth-debug).

### LINKED ISSUE #2637: Add all management API methods of Geth
### Description 

Implement all methods which are listed [here](https://github.com/ethereum/go-ethereum/wiki/Management-APIs).

#### Expected behavior

I can, for example, call the debug methods.

#### Actual behavior

Not all methods of the management API are supported.

### Versions 
* web3.js: latest

### PR DESCRIPTION
**Pull request by princesinha19**
## Description
I implemented all Geth node debugging management API's.
https://github.com/ethereum/go-ethereum/wiki/Management-APIs#list-of-management-apis

Fixes #2637 (Partially)

## Type of change
- [x] New feature 

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no warnings.
- [x] I have updated or added types for all modules I've changed
- [x] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran ```npm run test``` in the root folder with success and extended the tests if necessary.
- [ ] I ran ```npm run build``` in the root folder and tested it in the browser and with node.
- [ ] I ran ```npm run dtslint``` in the root folder and tested that all my types are correct
- [ ] I have tested my code on an ethereum test network.

### CODE REVIEW

**Review by nivida — APPROVED** [review]

**Review by nivida — APPROVED** [review]

**Comment by nivida** [comment]
I will write tomorrow the documentation for the Web3 Debug Module and merge your PR after that. Thanks, @princesinha19 for creating all these modules! :)

**Comment by princesinha19** [comment]
okay, @nivida. If you want me to write the documentation then, I can do that.

File: packages/web3-eth-debug/types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+    traceBlock(
+        blockRlp: string,
+        options?: {},
+        callback?: (error: Error, result: any) => void
+    ): Promise<any>;
+
+    traceBlockByNumber(
+        blockNumber: number,
+        options?: {},
```
Comment (joshstevens19) [inline packages/web3-eth-debug/types/index.d.ts]: please define this if you do not know do `any` `{}` will not let a `ts` developer pass in a dynamic options object here.

File: packages/web3-eth-debug/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        callback?: (error: Error, result: any) => void
+    ): Promise<any>;
+
+    traceBlockByHash(
+        transactionHash: string,
+        options?: {},
```
Comment (joshstevens19) [inline packages/web3-eth-debug/types/index.d.ts]: same here

File: packages/web3-eth-debug/types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        callback?: (error: Error, result: any) => void
+    ): Promise<any>;
+
+    traceBlockFromFile(
+        fileName: string,
+        options?: {},
```
Comment (joshstevens19) [inline packages/web3-eth-debug/types/index.d.ts]: same here

File: packages/web3-eth-debug/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        callback?: (error: Error, result: any) => void
+    ): Promise<any>;
+
+    traceTransaction(
+        transactionHash: string,
+        options?: {},
```
Comment (joshstevens19) [inline packages/web3-eth-debug/types/index.d.ts]: same here

File: packages/web3-eth-debug/types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+ * @file index.d.ts
+ * @author Prince Sinha <sinhaprince013@gmail.com>
+ * @date 2019
+ */
+
+import {provider} from 'web3-providers';
+import {AbstractWeb3Module, Web3ModuleOptions} from 'web3-core';
+import * as net from 'net';
+
+export class Debug extends AbstractWeb3Module {
```
Comment (joshstevens19) [inline packages/web3-eth-debug/types/index.d.ts:24]: In general you really cant type any of these methods at all? surely they have some sort of interface they follow? or is everything completely dynamic.. would like to get rid of all the `Promise<any>` if possible. `any` is used at the last resort and if its completely dynamic.

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: It's dynamic totally.

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: I will again try to create some interface. @joshstevens19. In documentation return types aren't defined so I need to test in geth all the methods.
some methods define log like:
```
> debug.startGoTrace('prince')

INFO [04-16|16:05:35.016] Go tracing started                       dump=prince
null
```
so, for these methods I can't decide any interface.

Comment (joshstevens19) [reply packages/web3-eth-debug/types/index.d.ts:24]: that just returns null though here, have you got ones with actual traces?

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: No @joshstevens19. I didn't tested with actual trace. I will trace and attach here.
But, for verbosity:-
```
debug.verbosity(5)

TRACE[04-16|17:33:24.267]                                          msg="<-readResp: response {\"jsonrpc\":\"2.0\",\"id\":12766,\"result\":null}"
null
> TRACE[04-16|17:33:24.326] >> FINDNODE/v4                           id=df5ff1b39bf70a61 addr=94.54.3.102:19485 err=nil
TRACE[04-16|17:33:24.687]                                          msg="sending {\"jsonrpc\":\"2.0\",\"id\":12767,\"method\":\"eth_getFilterChanges\",\"params\":[\"0xfdb0e2acdd8875c427b01ee199639c71\"]}"
TRACE[04-16|17:33:24.687]                                          msg="<-readResp: response {\"jsonrpc\":\"2.0\",\"id\":12767,\"result\":[]}"
TRACE[04-16|17:33:24.688]                                          msg="sending {\"jsonrpc\":\"2.0\",\"id\":12768,\"method\":\"eth_getFilterChanges\",\"params\":[\"0xa6e43d25960bd0c509deb830955b9031\"]}"
TRACE[04-16|17:33:24.688]                                          msg="<-readResp: response {\"jsonrpc\":\"2.0\",\"id\":12768,\"result\":[]}"
TRACE[04-16|17:33:24.826] Findnode failed                          id=df5ff1b39bf70a61 failcount=1 err="RPC timeout"
TRACE[04-16|17:33:24.826] Dial task done                           task="discovery lookup (16 results)"
TRACE[04-16|17:33:24.826] Skipping dial candidate                  id=3d1edcb0c6a7e329 addr=52.74.57.123:30303 err="recently dialed"
TRACE[04-16|17:33:24.826] Skipping dial candidate                  id=920a9fa96c2dd52a addr=106.243.159.4:30323 err="recently dialed"
TRACE[04-16|17:33:24.826] New dial task                            task="dyndial 92f3eb5656260131 142.93.231.24:30305"
TRACE[04-16|17:33:24.826] New dial task                            task="dyndial df4ff7d8c676272e 109.76.232.223:51307"
TRACE[04-16|17:33:24.826] New dial task                            task="dyndial df4f593d6c6e1a42 114.222.186.240:30917"
TRACE[04-16|17:33:24.826] New dial task                            task="dyndial df485a0b7f96321f 35.197.107.204:1102"
DEBUG[04-16|17:33:25.161] Adding p2p peer                          name=Geth/v1.8.23-stable-... addr=142.93.231.24:30305 peers=1
TRACE[04-16|17:33:25.161] connection set up                        id=92f3eb5656260131 addr=142.93.231.24:30305 conn=dyndial inbound=false
TRACE[04-16|17:33:25.161] Dial task done                           task="dyndial 92f3eb5656260131 142.93.231.24:30305"
TRACE[04-16|17:33:25.161] Starting protocol eth/63                 id=92f3eb5656260131 conn=dyndial
DEBUG[04-16|17:33:25.161] Ethereum peer connected                  id=92f3eb5656260131 conn=dyndial name=Geth/v1.8.23-stable-c9427004/linux-amd64/go1.11.5
TRACE[04-16|17:33:25.187]                                          msg="sending {\"jsonrpc\":\"2.0\",\"id\":12769,\"method\":\"eth_getFilterChanges\",\"params\":[\"0xfdb0e2acdd8875c427b01ee199639c71\"]}"
TRACE[04-16|17:33:25.188]                                          msg="<-readResp: response {\"jsonrpc\":\"2.0\",\"id\":12769,\"result\":[]}"
TRACE[04-16|17:33:25.188]                                          msg="sending {\"jsonrpc\":\"2.0\",\"id\":12770,\"method\":\"eth_getFilterChanges\",\"params\":[\"0xa6e43d25960bd0c509deb830955b9031\"]}"
TRACE[04-16|17:33:25.188]                                          msg="<-readResp: response {\"jsonrpc\":\"2.0\",\"id\":12770,\"result\":[]}"
TRACE[04-16|17:33:25.224] >> PING/v4                               id=921543b36c4d9b71 addr=188.166.92.230:28657 err=nil
DEBUG[04-16|17:33:25.282] Ethereum handshake failed                id=92f3eb5656260131 conn=dyndial err="Genesis block mismatch - ae113f53ede122d4 (!= a697c619ca29c7f5)"
TRACE[04-16|17:33:25.282] Protocol eth/63 failed                   id=92f3eb5656260131 conn=dyndial err="Genesis block mismatch - ae113f53ede122d4 (!= a697c619ca29c7f5)"
DEBUG[04-16|17:33:25.282] Removing p2p peer                        id=92f3eb5656260131 conn=dyndial duration=121.142ms peers=0 req=true err="subprotocol error"
TRACE[04-16|17:33:25.282] Skipping dial candidate                  id=29cca67d5db0fbb2 addr=5.1.83.226:30303     err="recently dialed"
TRACE[04-16|17:33:25.282] New dial task                            task="dyndial df4858660a97258a 54.169.160.137:30305"
TRACE[04-16|17:33:25.344] Dial error                               task="dyndial df4858660a97258a 54.169.160.137:30305"  err="dial tcp 54.169.160.137:30305: connect: connection refused"
TRACE[04-16|17:33:25.344] Dial task done                           task="dyndial df4858660a97258a 54.169.160.137:30305"
TRACE[04-16|17:33:25.345] Skipping dial candidate                  id=930cf49cd4de09a6 addr=52.16.188.185:30303  err="recently dialed"
TRACE[04-16|17:33:25.345] New dial task                            task="dyndial df4b6b6f980469eb 58.152.48.7:8085"
```
It continusly generates this so, I am worried that what could be the type for this.

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: same for debug.stacks():
```
debug.stacks()
goroutine 3759 [running]:
runtime/pprof.writeGoroutineStacks(0x110aa00, 0xc458ed3420, 0xc4447c9c80, 0xee95e0)
	/usr/lib/go-1.10/src/runtime/pprof/pprof.go:650 +0xa7
runtime/pprof.writeGoroutine(0x110aa00, 0xc458ed3420, 0x2, 0x1, 0x8)
	/usr/lib/go-1.10/src/runtime/pprof/pprof.go:639 +0x44
runtime/pprof.(*Profile).WriteTo(0x1813600, 0x110aa00, 0xc458ed3420, 0x2, 0x16, 0x20)
	/usr/lib/go-1.10/src/runtime/pprof/pprof.go:310 +0x3e4
github.com/ethereum/go-ethereum/internal/debug.(*HandlerT).Stacks(0xc42022a730, 0x0, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/internal/debug/api.go:195 +0x81
reflect.Value.call(0xc43214a420, 0xc420239d10, 0x13, 0xf8033c, 0x4, 0xc4323b2e60, 0x1, 0x1, 0x0, 0x3, ...)
	/usr/lib/go-1.10/src/reflect/value.go:447 +0x969
reflect.Value.Call(0xc43214a420, 0xc420239d10, 0x13, 0xc4203d0e60, 0x1, 0x1, 0x1014b18, 0xc4203d0de0, 0xc4203a3000)
	/usr/lib/go-1.10/src/reflect/value.go:308 +0xa4
github.com/ethereum/go-ethereum/rpc.(*Server).handle(0xc4320a0270, 0x1116d60, 0xc4320a00c0, 0x111c1e0, 0xc432130320, 0xc45a612e40, 0xc4203d0f38, 0x43f3b9, 0xc400000008)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/rpc/server.go:309 +0x676
github.com/ethereum/go-ethereum/rpc.(*Server).exec(0xc4320a0270, 0x1116d60, 0xc4320a00c0, 0x111c1e0, 0xc432130320, 0xc45a612e40)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/rpc/server.go:330 +0x1a6
github.com/ethereum/go-ethereum/rpc.(*Server).serveRequest.func2(0xc4323b6000, 0xc4320a0270, 0xc432152000, 0x111c1e0, 0xc432130320, 0xc4202392c0, 0x1, 0x1, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/rpc/server.go:204 +0xad
created by github.com/ethereum/go-ethereum/rpc.(*Server).serveRequest
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/rpc/server.go:199 +0x28c

goroutine 1 [chan receive]:
github.com/ethereum/go-ethereum/internal/jsre.(*JSRE).Do(...)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/internal/jsre/jsre.go:230
github.com/ethereum/go-ethereum/internal/jsre.(*JSRE).Evaluate(0xc4201cc4c0, 0xc444bb3d40, 0xf, 0x110c500, 0xc4200cc008, 0xc444bb3d40, 0xe)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/internal/jsre/jsre.go:311 +0x11c
github.com/ethereum/go-ethereum/console.(*Console).Evaluate(0xc432309030, 0xc444bb3d40, 0xf, 0x0, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/console/console.go:305 +0x87
github.com/ethereum/go-ethereum/console.(*Console).Interactive(0xc432309030)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/console/console.go:377 +0x429
main.localConsole(0xc4201242c0, 0x0, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/geth/consolecmd.go:109 +0x39a
github.com/ethereum/go-ethereum/cmd/utils.MigrateFlags.func1(0xc4201242c0, 0x0, 0xc4201242c0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/flags.go:1533 +0x103
github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0xdf6c80, 0xc4200bf720, 0xc4201242c0, 0xc4200ac600, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:490 +0xc8
github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.Command.Run(0xf83c07, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0xfaabf4, 0x2b, 0x0, ...)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli.v1/command.go:210 +0xa36
github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc4200e84e0, 0xc4200da000, 0x12, 0x12, 0x0, 0x0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:255 +0x6a0
main.main()
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/geth/main.go:256 +0x55

goroutine 36 [chan receive, 2 minutes]:
github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify.(*nonrecursiveTree).dispatch(0xc4200f6180, 0xc4200f60c0)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify/tree_nonrecursive.go:36 +0xb6
created by github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify.newNonrecursiveTree
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify/tree_nonrecursive.go:29 +0xf2

goroutine 37 [chan receive, 4 minutes]:
github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify.(*nonrecursiveTree).internal(0xc4200f6180, 0xc4200f6120)
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify/tree_nonrecursive.go:81 +0x7b
created by github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify.newNonrecursiveTree
	/build/ethereum-BhxhzZ/ethereum-1.8.26+build18465+bionic/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/rjeczalik/notify/tree_nonrecursive.go:30 +0x11e

goroutine 38 [syscall, 4 minutes]:
.
.
.
.
```

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: I can create interface for tracing transactions and blocks. debug.traceTransaction(), debug.traceBlockByNumber() etc..

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: Also for: debug.setHead()
```
debug.setHead('0x5')
WARN [04-16|17:44:09.724] Rewinding blockchain                     target=5
INFO [04-16|17:44:09.725] Loaded most recent local header          number=5 hash=4c470b…5ed1b6 td=655744 age=5m40s
INFO [04-16|17:44:09.725] Loaded most recent local full block      number=5 hash=4c470b…5ed1b6 td=655744 age=5m40s
INFO [04-16|17:44:09.725] Loaded most recent local fast block      number=5 hash=4c470b…5ed1b6 td=655744 age=5m40s
null
```

Comment (nivida) [reply packages/web3-eth-debug/types/index.d.ts:24]: I will merge this PR and improve the types before the next release. Thanks, @princesinha19 I really appreciate your work! 💪

Comment (princesinha19) [reply packages/web3-eth-debug/types/index.d.ts:24]: Ya will do. Thanks @nivida.

**Review by joshstevens19 — APPROVED** [review]
Top work 👍

**Comment by nivida** [comment]
@princesinha19 Could you implement the unresolved conversations here? Thanks!:)

**Comment by princesinha19** [comment]
Yes, I will do it today.

**Pull request merged**